packages/gestalt/src/Button.js

```diff
@@ -28,54 +29,65 @@ const DEFAULT_TEXT_COLORS = {
   white: 'darkGray',
 };
 
-const SIZE_NAME_TO_PIXEL = {
-  sm: 10,
-  md: 12,
-  lg: 12,
-};
-
-type Props = {|
-  accessibilityControls?: string,
-  accessibilityExpanded?: boolean,
-  accessibilityHaspopup?: boolean,
+type BaseButton = {|
   accessibilityLabel?: string,
   color?: 'gray' | 'red' | 'blue' | 'transparent' | 'white',
   disabled?: boolean,
-  iconEnd?: $Keys<typeof icons>,
   inline?: boolean,
-  name?: string,
-  onClick?: AbstractEventHandler<SyntheticMouseEvent<HTMLButtonElement>>,
-  selected?: boolean,
+  onClick?: AbstractEventHandler<
+    | SyntheticMouseEvent<HTMLButtonElement>
+    | SyntheticMouseEvent<HTMLAnchorElement>
+  >,
   size?: 'sm' | 'md' | 'lg',
   text: string,
   textColor?: 'white' | 'darkGray' | 'blue' | 'red',
-  type?: 'submit' | 'button',
 |};
 
+type ButtonType = {|
+  ...BaseButton,
+  accessibilityControls?: string,
+  accessibilityExpanded?: boolean,
+  accessibilityHaspopup?: boolean,
+  iconEnd?: $Keys<typeof icons>,
+  selected?: boolean,
+  type?: 'button',
+|};
+
+type SubmitButtonType = {|
+  ...BaseButton,
+  iconEnd?: $Keys<typeof icons>,
+  name?: string,
+  type: 'submit',
+|};
+
+type LinkButtonType = {|
+  ...BaseButton,
+  href: string,
+  rel?: 'none' | 'nofollow',
+  target?: null | 'self' | 'blank',
+  type: 'link',
+|};
+
+type unionProps = ButtonType | SubmitButtonType | LinkButtonType;
+
+type unionRefs = HTMLButtonElement | HTMLAnchorElement;
+
 const ButtonWithForwardRef: React$AbstractComponent<
-  Props,
-  HTMLButtonElement
-> = forwardRef<Props, HTMLButtonElement>(function Button(
-  props,
-  ref
-): Element<'button'> {
+  unionProps,
+  unionRefs
+> = forwardRef<unionProps, unionRefs>(function Button(props, ref): Node {
   const {
-    accessibilityControls,
-    accessibilityExpanded,
-    accessibilityHaspopup,
     accessibilityLabel,
     color = 'gray',
     disabled = false,
-    iconEnd,
     inline = false,
-    name,
     onClick,
     selected = false,
     size = 'md',
     text,
     textColor: textColorProp,
-    type = 'button',
```

Comment:
@AlbertCarreras could we leave `type = 'button'` as the default? Otherwise the default becomes "submit" and this PR becomes a breaking change which will most likely cause bugs in Pinterest's codebase.

![image](https://user-images.githubusercontent.com/127199/91466968-bdebf880-e844-11ea-81d8-bf91b3d66bb7.png)

See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type

Reply:
I probably missed this. The intention was to leave the button as default. The draft was more of the first exploratory implementation. But thanks for catching that.

Reply:
See line 231
![image](https://user-images.githubusercontent.com/10593890/91485596-efbe8880-e85f-11ea-9cd3-b93ccfd90174.png)

Reply:
DONE